### PR TITLE
Blocks: Fix `Stories` block rendering duplicate stories when globals are changed

### LINKED ISSUE #24751: [Bug]: Stories docblock duplicating stories on custom .MDX files
### Describe the bug

On a Storybook + Lit Web Components + Vite environment, when writing a custom .mdx file, the `<Stories/>` docblock duplicates the stories whenever a global option is ticked (e.g. background, grid, outline).

This happens when using custom MDX docs. When using autodocs, the stories are not duplicated, however, we have some components where we'd like to have more control over the docs, so using autodocs does not work for us.

### To Reproduce

[Here's a repo that can help replicate the issue.](https://github.com/xsu1010/storybook-lit-vite-docs-duplicating)

I also hosted it on Chromatic, [which you can check here.](https://654a6016d95b5229bf2bdeb1-daarfnmlph.chromatic.com/)

## Prerequisites:

- [pnpm](https://pnpm.io/) 

## Steps to reproduce:

1. Clone this repo:

```bash
git clone https://github.com/xsu1010/storybook-lit-vite-docs-duplicating.git
```

2. Install dependencies:

```bash
cd storybook-lit-vite-docs-duplicating && pnpm install
```

3. Run Storybook:

```bash
pnpm storybook
```

4. On the sidebar, you'll find a category named **Docs**. Click on any of the docs inside it and toggle through the global options on top (background, grid, outline).
   1. The Auto MDX doc will work just fine when toggling the global options.
   2. The Custom MDX docs will duplicate the stories when toggling the global options.

### System

```bash
Storybook Environment Info:

  System:
    OS: macOS 14.1
    CPU: (8) arm64 Apple M1 Pro
    Shell: 0.81.0 - /opt/homebrew/bin/nu
  Binaries:
    Node: 18.17.1 - ~/homebrew/bin/node
    Yarn: 1.22.19 - ~/homebrew/bin/yarn
    npm: 9.6.7 - ~/homebrew/bin/npm
    pnpm: 8.6.12 - ~/homebrew/bin/pnpm <----- active
  Browsers:
    Chrome: 119.0.6045.105
    Safari: 17.1
  npmPackages:
    @storybook/addon-essentials: ^7.5.3 => 7.5.3 
    @storybook/addon-links: ^7.5.3 => 7.5.3 
    @storybook/blocks: ^7.5.3 => 7.5.3 
    @storybook/web-components: ^7.5.3 => 7.5.3 
    @storybook/web-components-vite: ^7.5.3 => 7.5.3 
    chromatic: ^7.6.0 => 7.6.0 
    storybook: ^7.5.3 => 7.5.3
```

## Notes:
- This issue also happens with a static build (`pnpm build-storybook`).
- On any of the docs in the Docs category, I noticed that, when toggling through the global options, the page sort of refreshes fully,
sometimes throwing me to the top of the page.
  - However, this does not happen with the example docs (left them there for comparison). I'm not sure if this is related to the issue.

### PR DESCRIPTION
**Pull request by JReinhold**
Closes #24751

## What I did

This PR refactors `DocsContext` so it's more resilient in scenarios where multiple CSF files are being attached.

The cause of the linked issue was that rendering `<Meta />` would call `attachCSFFile`, and changing a global causes the whole docs view to re-render. This meant that `attachCSFFile` would be called again with the same CSF file, which would just append the stories to the existing array in `DocsContext` again and again.

Upon investigating this I concluded that `DocsContext` wasn't completely implemented in a way that handled multiple CSF files being attached at the same time. This _could theoretically_ result in the `primaryStory` being the primary of the _first attached_ CSF file, however a resolved `meta` or `component` would return the one from the _last attached_ CSF file. This has now been fixed in this PR - the first attached CSF file will always be considered the "primary" one to resolve from.

However I haven't figured out how to test this in practice, I can't set up a scenario where multiple _different_ CSF files are being attached to the same context. Based on the comment here, it suggest that this is a valid scenario though in non-manual-MDX files:

https://github.com/storybookjs/storybook/blob/332c593f8fb44fe62dabad1377da0517b42543f8/code/lib/preview-api/src/modules/preview-web/render/CsfDocsRender.ts#L106-L108

I tried having two different CSF files with the same title, and while that did collect all stories under the same component in the sidebar, it didn't make a difference in autodocs. @tmeasday any ideas on how I can test this out in a real Storybook?

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Open these docs, which is an MDX doc with `<Stories />`: https://630511d655df72125520f051-tprruuakio.chromatic.com/?path=/docs/addons-docs-docs2-button--metaof
2. Toggle any of the globals multiple times and see that the amount of stories stay the same.
3. Compare that with what's on `next`, where it duplicates the stories when globals are changed: https://next--630511d655df72125520f051.chromatic.com/?path=/docs/addons-docs-docs2-button--metaof

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: code/lib/preview-api/src/modules/preview-web/docs-context/DocsContext.ts [hunk 1]
```diff
@@ -71,13 +72,18 @@ export class DocsContext<TRenderer extends Renderer> implements DocsContextProps
     if (!this.exportsToCSFFile.has(csfFile.moduleExports)) {
       throw new Error('Cannot attach a CSF file that has not been referenced');
     }
+    if (this.attachedCSFFiles.has(csfFile)) {
+      // this CSF file is already attached, don't do anything
+      return;
+    }
```
Comment (JReinhold) [inline code/lib/preview-api/src/modules/preview-web/docs-context/DocsContext.ts:78]: This is the actual fix for the issue, if we actually don't think supporting multiple attached CSF file is a valid use case then this PR can be reduced to these lines.

**Comment by tmeasday** [comment]
> I tried having two different CSF files with the same title, and while that did collect all stories under the same component in the sidebar, it didn't make a difference in autodocs. @tmeasday any ideas on how I can test this out in a real Storybook?

@JReinhold did the index end up with the `storiesImport` containing a second CSF @JReinhold ?

According to this test it should: https://github.com/storybookjs/storybook/blob/82504d634c86ca426894703be3166201076bcb02/code/lib/core-server/src/utils/StoryIndexGenerator.test.ts#L686

In which case I would have expected the line referenced by that comment to run, which would attach the second CSF file's stories to the docs context, and they'd show up in autodocs. I guess that's not happening, I would have to debug through why. Maybe we can do it together?

**Review by tmeasday — COMMENTED** [review]
I think the changes look good apart from one nit.

I think maybe we should try and add some sandbox stories for this situation and maybe an E2E test that checks the behaviour?

File: code/lib/preview-api/src/modules/preview-web/docs-context/DocsContext.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ import type { StoryStore } from '../../store';
 import type { DocsContextProps } from './DocsContextProps';
 
 export class DocsContext<TRenderer extends Renderer> implements DocsContextProps<TRenderer> {
-  private componentStoriesValue: PreparedStory<TRenderer>[];
+  private componentStoriesValue: Set<PreparedStory<TRenderer>>;
```
Comment (tmeasday) [inline code/lib/preview-api/src/modules/preview-web/docs-context/DocsContext.ts:20]: My one nit seems to have dissappeared. It was that I don't understand why this needed to change to a set, that seems to unnecessarily complicate a value that is an array at the end of the day.

Comment (JReinhold) [reply code/lib/preview-api/src/modules/preview-web/docs-context/DocsContext.ts:20]: The main reason was to ensure we don't add  duplicates of the same stories, but we are also ensuring that with other means, so I'm okay with reverting this.

**Comment by JReinhold** [comment]
@tmeasday I don't know what happened, but when adding the E2E tests it works as expected, stories from multiple CSF files are merged in autodocs. I must have gotten my test-setup wrong. 

It doesn't work in MDX but that's to be expected.

**Pull request merged**